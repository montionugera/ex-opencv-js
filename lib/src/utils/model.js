import * as fp from 'lodash/fp'

import { validateModel, getModelErrorMessages, isNotEmpty } from 'validation'

const getErrorMessages = ({ validate: config, field }) => ({ session }) =>
  fp.map(({ validate, message }) => ({
    message,
    isValid: validate(fp.get(field, session), session),
  }))(config)

const isValid = ({ validate: config }) => (...args) =>
  fp.all(({ validate }) => validate(...args), config)

export const createModel = fp.pipe(
  fp.traverse(
    fp.pipe(
      fp.when(fp.has('path'), (config, path, obj) =>
        fp.set('field', `${obj.base}.${config.path}`, config)),
      fp.when(fp.isMatch({ required: true }), (config) =>
        fp.update(
          'validate',
          fp.pipe(
            fp.defaultTo([]),
            fp.prepend({
              message: 'validation:error.empty',
              validate: isNotEmpty,
            })
          ),
          config
        )),
      fp.when(fp.has('validate'), (config) => fp.set('isValid', isValid(config), config)),
      fp.when(fp.has('validate'), (config) =>
        fp.set(
          'getErrorMessages',
          fp.pipe(
            getErrorMessages(config),
            fp.filter(fp.isMatch({ isValid: false })),
            fp.map(fp.get('message'))
          ),
          config
        )),
      fp.when(fp.has('validate'), (config) =>
        fp.set(
          'validate',
          fp.pipe(
            getErrorMessages(config),
            fp.all(fp.isMatch({ isValid: true }))
          ),
          config
        ))
    )
  ),
  (config) => fp.set('validate', validateModel(config.fields), config),
  (config) => fp.set('getErrorMessages', getModelErrorMessages(config.fields), config)
)
