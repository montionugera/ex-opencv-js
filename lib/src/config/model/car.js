import * as fp from 'lodash/fp'

import {
  isName,
  isNotEmpty,
  isDate,
  isNumeric,
  isThaiNationalId,
} from 'validation'

import { ONCE, MONTHLY, YEARLY } from 'constants/payment'
import { createModel } from 'utils/model'
import getMasterData from 'services/masterdata'
import callMiddleware from 'services/middleware'
import i18n from 'i18n'

export const car = createModel({
  base: 'data.car',
  fields: {
    brand: {
      path: 'brand',
      label: 'car:car.brand.label',
      placeholder: 'car:car.brand.placeholder',
      type: 'select',
      required: true,
      dependencies: {
        suggestion: ({ session, token }) =>
          getMasterData('masterdata.car.suggestBrand', {
            data: {
              brand: fp.get(car.fields.brand.field, session),
            },
            token,
            session,
          }),
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listBrand', {
            token,
            session,
          }),
      },
    },
    model: {
      path: 'model',
      label: 'car:car.model.label',
      placeholder: 'car:car.model.placeholder',
      type: 'select',
      required: true,
      dependencies: {
        context: (session) =>
          fp.applySpec({
            brand: fp.get(car.fields.brand.field),
          })(session),
        suggestion: ({ session, token }) =>
          getMasterData('masterdata.car.suggestModel', {
            data: {
              brand: fp.get(car.fields.brand.field, session),
              model: fp.get(car.fields.model.field, session),
            },
            token,
            session,
          }),
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listModel', {
            data: {
              brand: fp.get(car.fields.brand.field, session),
            },
            token,
            session,
          }),
      },
    },
    year: {
      path: 'year',
      label: 'car:car.year.label',
      placeholder: 'car:car.year.placeholder',
      type: 'select',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
      dependencies: {
        context: (session) =>
          fp.applySpec({
            brand: fp.get(car.fields.brand.field),
            model: fp.get(car.fields.model.field),
          })(session),
        suggestion: ({ session, token }) =>
          getMasterData('masterdata.car.suggestYear', {
            data: {
              brand: fp.get(car.fields.brand.field, session),
              model: fp.get(car.fields.model.field, session),
              year: fp.get(car.fields.year.field, session),
            },
            token,
            session,
          }),
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listYear', {
            data: {
              brand: fp.get(car.fields.brand.field, session),
              model: fp.get(car.fields.model.field, session),
            },
            token,
            session,
          }),
      },
    },
    province: {
      path: 'province',
      label: 'car:car.province.label',
      placeholder: 'car:car.province.placeholder',
      type: 'select',
      required: true,
      dependencies: {
        suggestion: ({ session, token }) =>
          getMasterData('masterdata.car.suggestProvince', {
            data: {
              language: session.language,
              province: fp.get(car.fields.province.field, session),
            },
            token,
            session,
          }),
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listProvince', {
            data: {
              language: session.language,
            },
            token,
            session,
          }),
      },
    },
    id: {
      path: 'id',
      label: 'car:car.id.label',
      placeholder: 'car:car.id.placeholder',
      type: 'text',
      required: true,
    },
    chassis_number: {
      path: 'chassis_number',
      label: 'car:car.chassis_number.label',
      placeholder: 'car:car.chassis_number.placeholder',
      type: 'text',
      required: true,
    },
  },
})

export const insured = createModel({
  base: 'data.insured',
  fields: {
    name: {
      path: 'name',
      label: 'car:insured.name.label',
      placeholder: 'car:insured.name.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_name',
          validate: isName,
        },
      ],
    },
    national_id: {
      path: 'national_id',
      label: 'car:insured.national_id.label',
      placeholder: 'car:insured.national_id.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
        // {
        //   message: 'validation:error.invalid_thai_national_id',
        //   validate: isThaiNationalId,
        // },
      ],
    },
    dob: {
      path: 'dob',
      label: 'car:insured.dob.label',
      placeholder: 'car:insured.dob.placeholder',
      type: 'date',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_date',
          validate: isDate,
        },
      ],
    },
    delivery_location: {
      path: 'delivery_location',
      label: 'car:insured.delivery_location.label',
      placeholder: 'car:insured.delivery_location.placeholder',
      type: 'textarea',
      required: true,
    },
    job: {
      path: 'job',
      label: 'car:insured.job.label',
      placeholder: 'car:insured.job.placeholder',
      type: 'text',
    },
  },
})

export const attachment = createModel({
  base: 'attachment',
  fields: {
    car_front: {
      path: 'car_front',
      label: 'attachment:car_front',
      type: 'image',
      required: true,
    },
    car_back: {
      path: 'car_back',
      label: 'attachment:car_back',
      type: 'image',
      required: true,
    },
    car_left: {
      path: 'car_left',
      label: 'attachment:car_left',
      type: 'image',
      required: true,
    },
    car_right: {
      path: 'car_right',
      label: 'attachment:car_right',
      type: 'image',
      required: true,
    },
    car_registration: {
      path: 'car_registration',
      label: 'attachment:car_registration',
      type: 'image',
      required: true,
    },
    national_id: {
      path: 'national_id',
      label: 'attachment:national_id',
      type: 'image',
      required: true,
    },
  },
})

export const carProduct = createModel({
  base: 'data.product.car',
  dependencies: {
    suggestion: ({ session, token, data }) =>
      getMasterData('masterdata.car.suggestProduct', {
        data,
        token,
        session,
      }),
    list: async ({ session, token, action }) => {
      const { productFilter } = action.args[0]

      const result = await getMasterData('masterdata.car.suggestProduct', {
        data: productFilter,
        token,
        session,
      })

      return result
    },
  },
  enable: fp.allPass([car.validate]),
  fields: {
    id: {
      path: 'id',
      label: 'car:product.id',
      required: true,
      dependencies: {
        context: (session) => fp.get(carProduct.fields.id.field, session),
        list: async ({ token, session, action }) => {
          const result = await getMasterData('masterdata.car.getProduct', {
            data: {
              id:
                fp.get('args[0]', action) ||
                fp.get(carProduct.fields.id.field, session),
            },
            token,
            session,
          })

          return result
        },
      },
    },
    type: {
      path: 'type',
      label: 'car:product.type',
      dependencies: {
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listProductType', {
            token,
            session,
          }),
      },
    },
    price: {
      path: 'price',
      label: 'car:product.price',
    },
    topup: {
      path: 'topup',
      label: 'car:product.topup',
    },
    commission_percent: {
      path: 'commission_percent',
      label: 'car:product.commission_percent',
    },
    brand: {
      path: 'brand',
      label: 'car:product.brand',
      dependencies: {
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listProductBrand', {
            token,
            session,
          }),
      },
    },
    repair_type: {
      path: 'car_coverage.repair_type',
      label: 'car:product.repair_type',
      dependencies: {
        list: ({ session, token }) =>
          getMasterData('masterdata.car.listProductRepairType', {
            token,
            session,
          }),
      },
    },
  },
})

export const addition = createModel({
  base: 'data.product.car',
  fields: {
    has_cctv: {
      path: 'addition.has_cctv',
      label: 'car:addition.has_cctv.text',
      type: 'checkbox',
    },
    is_buy_car_act: {
      path: 'addition.is_buy_car_act',
      label: 'car:addition.buy_car_act.text',
      type: 'checkbox',
    },
    has_first_driver: {
      path: 'addition.has_first_driver',
      label: 'car:addition.first_driver',
      type: 'checkbox',
    },
    first_driver_name: {
      path: 'addition.first_driver.name',
      label: 'car:addition.first_driver_name.label',
      placeholder: 'car:addition.first_driver_name.placeholder',
      enable: ({ session }) =>
        fp.get(addition.fields.has_first_driver.field, session) === true,
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_name',
          validate: isName,
        },
      ],
    },
    first_driver_dob: {
      path: 'addition.first_driver.dob',
      label: 'car:addition.first_driver_dob.label',
      placeholder: 'car:addition.first_driver_dob.placeholder',
      enable: ({ session }) =>
        fp.get(addition.fields.has_first_driver.field, session) === true,
      type: 'date',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_date',
          validate: isDate,
        },
      ],
    },
    first_driver_job: {
      path: 'addition.first_driver.job',
      label: 'car:addition.first_driver_job.label',
      placeholder: 'car:addition.first_driver_job.placeholder',
      enable: ({ session }) =>
        fp.get(addition.fields.has_first_driver.field, session) === true,
      type: 'text',
    },
    has_second_driver: {
      path: 'addition.has_second_driver',
      label: 'car:addition.second_driver',
      type: 'checkbox',
    },
    second_driver_name: {
      path: 'addition.second_driver.name',
      label: 'car:addition.second_driver_name.label',
      placeholder: 'car:addition.second_driver_name.placeholder',
      enable: ({ session }) =>
        fp.get(addition.fields.has_second_driver.field, session) === true,
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_name',
          validate: isName,
        },
      ],
    },
    second_driver_dob: {
      path: 'addition.second_driver.dob',
      label: 'car:addition.second_driver_dob.label',
      placeholder: 'car:addition.second_driver_dob.placeholder',
      enable: ({ session }) =>
        fp.get(addition.fields.has_second_driver.field, session) === true,
      type: 'date',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_date',
          validate: isDate,
        },
      ],
    },
    second_driver_job: {
      path: 'addition.second_driver.job',
      label: 'car:addition.second_driver_job.label',
      placeholder: 'car:addition.second_driver_job.placeholder',
      enable: ({ session }) =>
        fp.get(addition.fields.has_second_driver.field, session) === true,
      type: 'text',
    },
  },
})

export const application = createModel({
  base: 'data.application.car',
  dependencies: {
    list: async ({ session, token }) => {
      const result = await callMiddleware('application.car.generatePDF', {
        token,
        data: {
          sessionId: session.id,
          type: 'image/jpeg',
        },
        responseType: 'blob',
      })

      return URL.createObjectURL(new Blob([result]))
    },
  },
  fields: {
    status: {
      path: 'status',
    },
    type: {
      path: 'type',
    },
  },
})

export const payment = createModel({
  base: 'data.payment.car',
  dependencies: {
    list: ({ session, token }) =>
      Promise.all(
        fp.pipe(
          fp.get(payment.fields.transactions.field),
          fp.map((transactionId) =>
            callMiddleware('payment.get', {
              token,
              data: {
                sessionId: session.id,
                transactionId,
              },
            })
          )
        )(session)
      ),
  },
  enable: fp.allPass([
    insured.validate,
    car.validate,
    carProduct.validate,
    addition.validate,
    attachment.validate,
  ]),
  fields: {
    transactions: {
      path: 'transactions',
    },
    mode: {
      dependencies: {
        list: async ({ session }) => {
          const { language } = session
          return [
            {
              label: i18n().getFixedT(language)('payment:mode.once'),
              value: ONCE,
            },
            {
              label: i18n().getFixedT(language)('payment:mode.monthly'),
              value: MONTHLY,
            },
            {
              label: i18n().getFixedT(language)('payment:mode.yearly'),
              value: YEARLY,
            },
          ]
        },
      },
      path: 'mode',
    },
  },
})
