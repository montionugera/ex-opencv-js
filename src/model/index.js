export const base = {
  car_registration_upper: require('./base/car-registration-upper'),
  national_id: require('./base/national-id'),
}

export const mapping = {
  car_registration_upper: require('./mapping/car-registration-upper'),
  national_id: require('./mapping/national-id'),
}
