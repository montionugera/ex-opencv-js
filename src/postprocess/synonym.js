import * as fp from 'lodash/fp'

import fuzz from 'fuzzball'

const SYNONYM_RATIO = 0.7

export const findSynonym = (word, list) => {
  if (!word || !fp.isString(word)) return undefined

  const synonym = fuzz.extract(word, list, {
    scorer: fuzz.partial_ratio,
    limit: 1,
    cutoff: SYNONYM_RATIO * 100,
  })

  return fp.pipe(fp.flatten, fp.first)(synonym)
}

export default (fields, sceneList) => {
  const result = fp.map((data) => {
    const field = fp.get('field', data)

    const synonym = fp.get([field, 'synonym'], fields)

    if (synonym) {
      const description = fp.get('description', data)

      const result = findSynonym(description, synonym)
      if (result) {
        return fp.set('description', result, data)
      }
    }

    return data
  })(sceneList)

  return result
}
