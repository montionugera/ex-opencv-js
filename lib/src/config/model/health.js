import * as fp from 'lodash/fp'

import { isName, isDate, isNumeric, isThaiNationalId } from 'validation'

import { ONCE, MONTHLY, YEARLY } from 'constants/payment'
import { createModel } from 'utils/model'
import getMasterData from 'services/masterdata'
import callMiddleware from 'services/middleware'
import i18n from 'i18n'

export const healthCancer = createModel({
  base: 'data.health.cancer',
  fields: {
    has_excepted_cancer: {
      hidden: true,
      path: 'has_excepted_cancer',
      label: 'health:health.cancer.has_excepted_cancer',
      type: 'checkbox_button',
      required: true,
    },
    excepted_cancer: {
      hidden: true,
      path: 'excepted_cancer.name',
      label: 'health:health.cancer.excepted_cancer.label',
      placeholder: 'health:health.cancer.excepted_cancer.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) =>
        fp.get(healthCancer.fields.has_excepted_cancer.field, session) === true,
    },
    has_cancer: {
      hidden: true,
      path: 'has_cancer',
      label: 'health:health.cancer.has_cancer',
      type: 'checkbox_button',
      required: true,
    },
    cancer: {
      hidden: true,
      path: 'cancer.name',
      label: 'health:health.cancer.cancer.label',
      placeholder: 'health:health.cancer.cancer.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) => fp.get(healthCancer.fields.has_cancer.field, session) === true,
    },
    smoking_history: {
      hidden: true,
      type: 'label',
      label: 'health:health.cancer.smoking_history',
    },
    is_smoking: {
      hidden: true,
      path: 'is_smoking',
      label: 'health:health.cancer.is_smoking',
      type: 'checkbox_button',
      required: true,
    },
    smoking_amount: {
      hidden: true,
      path: 'smoking.amount',
      label: 'health:health.cancer.smoking.amount.label',
      placeholder: 'health:health.cancer.smoking.amount.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
      enable: ({ session }) => fp.get(healthCancer.fields.is_smoking.field, session) === true,
    },
    smoking_start: {
      hidden: true,
      path: 'smoking.start',
      label: 'health:health.cancer.smoking.start.label',
      placeholder: 'health:health.cancer.smoking.start.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
      enable: ({ session }) => fp.get(healthCancer.fields.is_smoking.field, session) === true,
    },
    smoking_end: {
      hidden: true,
      path: 'smoking.end',
      label: 'health:health.cancer.smoking.end.label',
      placeholder: 'health:health.cancer.smoking.end.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
      enable: ({ session }) => fp.get(healthCancer.fields.is_smoking.field, session) === true,
    },
    is_still_smoking: {
      hidden: true,
      path: 'is_still_smoking',
      label: 'health:health.cancer.is_still_smoking',
      type: 'checkbox_button',
      required: true,
    },
    currently_smoking_amount: {
      hidden: true,
      path: 'currently_smoking.amount',
      label: 'health:health.cancer.currently_smoking.amount.label',
      placeholder: 'health:health.cancer.currently_smoking.amount.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
      enable: ({ session }) => fp.get(healthCancer.fields.is_still_smoking.field, session) === true,
    },
    disease: {
      hidden: true,
      type: 'label',
      label: 'health:health.cancer.disease',
    },
    has_disease_a: {
      hidden: true,
      path: 'has_disease_a',
      label: 'health:health.cancer.has_disease_a',
      type: 'checkbox_button',
      required: true,
    },
    disease_a_name: {
      hidden: true,
      path: 'disease.a.name',
      label: 'health:health.cancer.disease_a.name.label',
      placeholder: 'health:health.cancer.disease_a.name.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) => fp.get(healthCancer.fields.has_disease_a.field, session) === true,
    },
    has_disease_b: {
      hidden: true,
      path: 'has_disease_b',
      label: 'health:health.cancer.has_disease_b',
      type: 'checkbox_button',
      required: true,
    },
    disease_b_type: {
      hidden: true,
      path: 'disease.b.type',
      label: 'health:health.cancer.disease_b.type.label',
      placeholder: 'health:health.cancer.disease_b.type.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) => fp.get(healthCancer.fields.has_disease_b.field, session) === true,
    },
    disease_b_organ: {
      hidden: true,
      path: 'disease.b.organ',
      label: 'health:health.cancer.disease_b.organ.label',
      placeholder: 'health:health.cancer.disease_b.organ.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) => fp.get(healthCancer.fields.has_disease_b.field, session) === true,
    },
    has_disease_c: {
      hidden: true,
      path: 'has_disease_c',
      label: 'health:health.cancer.has_disease_c',
      type: 'checkbox_button',
      required: true,
    },
    disease_c_type: {
      hidden: true,
      path: 'disease.c.type',
      label: 'health:health.cancer.disease_c.type.label',
      placeholder: 'health:health.cancer.disease_c.type.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) => fp.get(healthCancer.fields.has_disease_c.field, session) === true,
    },
    has_insurance_company: {
      hidden: true,
      path: 'has_insurance_company',
      label: 'health:health.cancer.has_insurance_company',
      type: 'checkbox_button',
      required: true,
    },
    insurance_company_name: {
      hidden: true,
      path: 'insurance_company.name',
      label: 'health:health.cancer.insurance_company.name.label',
      placeholder: 'health:health.cancer.insurance_company.name.placeholder',
      type: 'text',
      required: true,
      enable: ({ session }) =>
        fp.get(healthCancer.fields.has_insurance_company.field, session) === true,
    },
    insurance_company_amount: {
      hidden: true,
      path: 'insurance_company.amount',
      label: 'health:health.cancer.insurance_company.amount.label',
      placeholder: 'health:health.cancer.insurance_company.amount.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
      enable: ({ session }) =>
        fp.get(healthCancer.fields.has_insurance_company.field, session) === true,
    },
  },
})

export const insured = createModel({
  base: 'data.insured',
  fields: {
    name: {
      path: 'name',
      label: 'health:insured.name.label',
      placeholder: 'health:insured.name.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_name',
          validate: isName,
        },
      ],
    },
    gender: {
      path: 'gender',
      label: 'health:insured.gender.label',
      placeholder: 'health:insured.gender.placeholder',
      type: 'select',
      required: true,
      dependencies: {
        list: ({ session, token }) =>
          getMasterData('masterdata.health.listGender', {
            token,
            session,
          }),
      },
    },
    dob: {
      path: 'dob',
      label: 'health:insured.dob.label',
      placeholder: 'health:insured.dob.placeholder',
      type: 'date',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_date',
          validate: isDate,
        },
      ],
    },
    national_id: {
      path: 'national_id',
      label: 'health:insured.national_id.label',
      placeholder: 'health:insured.national_id.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
        // {
        //   message: 'validation:error.invalid_thai_national_id',
        //   validate: isThaiNationalId,
        // },
      ],
    },
    current_location: {
      path: 'current_location',
      label: 'health:insured.current_location.label',
      placeholder: 'health:insured.current_location.placeholder',
      type: 'textarea',
      required: true,
    },
    work_location: {
      path: 'work_location',
      label: 'health:insured.work_location.label',
      placeholder: 'health:insured.work_location.placeholder',
      type: 'textarea',
      required: true,
    },
    job: {
      path: 'job',
      label: 'health:insured.job.label',
      placeholder: 'health:insured.job.placeholder',
      type: 'text',
    },
    income_per_month: {
      path: 'income_per_month',
      label: 'health:insured.income_per_month.label',
      placeholder: 'health:insured.income_per_month.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
    },
    height: {
      path: 'height',
      label: 'health:insured.height.label',
      placeholder: 'health:insured.height.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
    },
    weight: {
      path: 'weight',
      label: 'health:insured.weight.label',
      placeholder: 'health:insured.weight.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_number',
          validate: isNumeric,
        },
      ],
    },
  },
})

export const beneficiary = createModel({
  base: 'data.beneficiary',
  fields: {
    name: {
      path: 'name',
      label: 'health:beneficiary.name.label',
      placeholder: 'health:beneficiary.name.placeholder',
      type: 'text',
      required: true,
      validate: [
        {
          message: 'validation:error.invalid_name',
          validate: isName,
        },
      ],
    },
    relationship: {
      path: 'relationship',
      label: 'health:beneficiary.relationship.label',
      placeholder: 'health:beneficiary.relationship.placeholder',
      type: 'text',
      required: true,
    },
    location: {
      path: 'location',
      label: 'health:beneficiary.location.label',
      placeholder: 'health:beneficiary.location.placeholder',
      type: 'textarea',
      required: true,
    },
  },
})

export const attachment = createModel({
  base: 'attachment',
  fields: {
    national_id: {
      path: 'national_id',
      label: 'attachment:national_id',
      type: 'image',
      required: true,
    },
  },
})

export const healthProduct = createModel({
  base: 'data.product.health',
  dependencies: {
    suggestion: ({ session, token, data }) =>
      getMasterData('masterdata.health.suggestProduct', {
        data,
        token,
        session,
      }),
    list: async ({ session, token, action }) => {
      const { productFilter } = action.args[0]

      const result = await getMasterData('masterdata.health.suggestProduct', {
        data: productFilter,
        token,
        session,
      })

      return result
    },
  },
  enable: fp.allPass([healthCancer.validate]),
  fields: {
    id: {
      path: 'id',
      label: 'health:product.id',
      required: true,
      // dependencies: {
      //   list: async ({ token, session, action }) => {
      //     const result = await getMasterData('masterdata.health.getProduct', {
      //       data: {
      //         id: fp.get('args[0]', action) || fp.get(healthProduct.fields.id.field, session),
      //       },
      //       token,
      //       session,
      //     })

      //     return result
      //   },
      // },
    },
    type: {
      path: 'type',
      label: 'health:product.type',
      // dependencies: {
      //   list: ({ session, token }) =>
      //     getMasterData('masterdata.health.listProductType', {
      //       token,
      //       session,
      //     }),
      // },
    },
    plan_id: {
      path: 'plan_id',
      label: 'health:product.plan_id',
      required: true,
      dependencies: {
        context: (session) =>
          fp.applySpec({
            id: fp.get(healthProduct.fields.id.field),
            plan_id: fp.get(healthProduct.fields.plan_id.field),
          })(session),
        list: async ({ token, session, action }) => {
          const result = await getMasterData('masterdata.health.getProductPlan', {
            data: {
              id: fp.get('args[0]', action) || fp.get(healthProduct.fields.id.field, session),
              planId:
                fp.get('args[1]', action) || fp.get(healthProduct.fields.plan_id.field, session),
            },
            token,
            session,
          })

          return result
        },
      },
    },
  },
})

export const application = createModel({
  base: 'data.application.health',
  // dependencies: {
  //   list: async ({ session, token }) => {
  //     const result = await callMiddleware('application.health.generatePDF', {
  //       token,
  //       data: {
  //         sessionId: session.id,
  //         type: 'image/jpeg',
  //       },
  //       responseType: 'blob',
  //     })

  //     return URL.createObjectURL(new Blob([result]))
  //   },
  // },
  fields: {
    status: {
      path: 'status',
    },
    type: {
      path: 'type',
    },
  },
})

export const payment = createModel({
  base: 'data.payment.health',
  dependencies: {
    list: ({ session, token }) =>
      Promise.all(
        fp.pipe(
          fp.get(payment.fields.transactions.field),
          fp.map((transactionId) =>
            callMiddleware('payment.get', {
              token,
              data: {
                sessionId: session.id,
                transactionId,
              },
            }))
        )(session)
      ),
  },
  enable: fp.allPass([
    insured.validate,
    beneficiary.validate,
    healthCancer.validate,
    healthProduct.validate,
    attachment.validate,
  ]),
  fields: {
    transactions: {
      path: 'transactions',
    },
    mode: {
      dependencies: {
        list: async ({ session }) => {
          const { language } = session
          return [
            {
              label: i18n().getFixedT(language)('payment:mode.once'),
              value: ONCE,
            },
            {
              label: i18n().getFixedT(language)('payment:mode.monthly'),
              value: MONTHLY,
            },
            {
              label: i18n().getFixedT(language)('payment:mode.yearly'),
              value: YEARLY,
            },
          ]
        },
      },
      path: 'mode',
    },
  },
})
