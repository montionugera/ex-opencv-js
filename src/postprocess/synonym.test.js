import * as model from 'model'
import synonym from './synonym'

describe('synonym', () => {
  it('return correctly', () => {
    const sceneList = [
      {
        boundingPoly: {
          vertices: [
            {
              x: 884,
              y: 640,
            },
            {
              x: 1434,
              y: 656,
            },
            {
              x: 1449,
              y: 718,
            },
            {
              x: 874,
              y: 700,
            },
          ],
        },
        description: 'กรงเทพบหานคร',
        confidence: 0.6,
        field: 'vehicle_license_province',
      },
      {
        boundingPoly: {
          vertices: [
            {
              x: 884,
              y: 640,
            },
            {
              x: 1434,
              y: 656,
            },
            {
              x: 1449,
              y: 718,
            },
            {
              x: 874,
              y: 700,
            },
          ],
        },
        description: 'กรง',
        confidence: 0.3,
        field: 'vehicle_license_province',
      },
    ]

    const type = 'car_registration_upper'
    const { fields } = model.mapping[type]

    const result = synonym(fields, sceneList)
    expect(result).toMatchSnapshot()
  })
})
