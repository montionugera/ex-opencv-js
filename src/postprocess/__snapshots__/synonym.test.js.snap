// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`synonym return correctly 1`] = `
Array [
  Object {
    "boundingPoly": Object {
      "vertices": Array [
        Object {
          "x": 884,
          "y": 640,
        },
        Object {
          "x": 1434,
          "y": 656,
        },
        Object {
          "x": 1449,
          "y": 718,
        },
        Object {
          "x": 874,
          "y": 700,
        },
      ],
    },
    "confidence": 0.6,
    "description": "กรุงเทพมหานคร",
    "field": "vehicle_license_province",
  },
  Object {
    "boundingPoly": Object {
      "vertices": Array [
        Object {
          "x": 884,
          "y": 640,
        },
        Object {
          "x": 1434,
          "y": 656,
        },
        Object {
          "x": 1449,
          "y": 718,
        },
        Object {
          "x": 874,
          "y": 700,
        },
      ],
    },
    "confidence": 0.3,
    "description": "กรง",
    "field": "vehicle_license_province",
  },
]
`;
